examples/search_augmentation.ipynb

```diff
@@ -0,0 +1,472 @@
+    "    print()\n"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "There is a lot of noise in these results, let's now use re-ranking and the completions model to synthesize a good final answer out of all these news articles.\n",
+    "\n",
+    "# 2. Re-rank\n",
+    "\n",
+    "Let's first generate a hypothetical answer using the completions API.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "\"Recently in San Francisco, the city has implemented a new COVID-19 vaccine mandate for indoor activities and events. Additionally, the Golden State Warriors have started their NBA season, playing games at the Chase Center arena. There have also been ongoing discussions about the city's affordable housing crisis and the need for more solutions to address it.\""
+      ]
+     },
+     "execution_count": 5,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "HA_PROMPT = f\"\"\"\n",
+    "Generate a hypothetical answer to the user's question. This answer which will be used to rank search results. Pretend you have all the information you need to answer.\n",
+    "\n",
+    "User question: {USER_QUESTION}\n",
+    "\n",
+    "Format: {{\"hypotheticalAnswer\": \"hypothetical answer text\"}}\n",
+    "\"\"\"\n",
+    "\n",
+    "hypothetical_answer = json_gpt(HA_PROMPT)[\"hypotheticalAnswer\"]\n",
+    "\n",
+    "hypothetical_answer\n"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, let's generate embeddings for the search results and the hypothetical answer.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "[0.7181421473937228,\n",
+       " 0.7047313247008025,\n",
+       " 0.7833773244357916,\n",
+       " 0.7229598385205838,\n",
+       " 0.7561981577536391,\n",
+       " 0.7978369364374425,\n",
+       " 0.765459048492688,\n",
+       " 0.768682248268667,\n",
+       " 0.7719928527938629,\n",
+       " 0.7750573135338864]"
+      ]
+     },
+     "execution_count": 6,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "hypothetical_answer_embedding = embedding(hypothetical_answer)\n",
+    "article_embeddings = embedding(\n",
+    "    [\n",
+    "        f\"{article['title']} {article['description']} {article['content'][0:100]}\"\n",
+    "        for article in articles\n",
+    "    ]\n",
+    ")\n",
+    "\n",
+    "# Calculate cosine similarity\n",
+    "cosine_similarities = []\n",
+    "for article_embedding in article_embeddings:\n",
+    "    cosine_similarities.append(\n",
+    "        dot(hypothetical_answer_embedding, article_embedding)[0])\n",
```

Comment:
also worth commenting why dot product works as cosine similarity, since that's not generally true